genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GrpcAgentHeartBeatServiceImpl.java

```diff
@@ -0,0 +1,178 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl.grpc;
+
+import com.netflix.genie.agent.execution.services.AgentHeartBeatService;
+import com.netflix.genie.proto.AgentHeartBeat;
+import com.netflix.genie.proto.HeartBeatServiceGrpc;
+import com.netflix.genie.proto.ServerHeartBeat;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.annotation.Validated;
+
+import javax.validation.constraints.NotBlank;
+import java.time.Instant;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * gRPC implementation of AgentHeartBeatService.
+ * Sends heartbeats to the server.
+ * Transparently handles disconnections and stream errors by establishing a new stream.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+@Validated
+class GrpcAgentHeartBeatServiceImpl implements AgentHeartBeatService {
+
+    private static final long HEART_BEAT_PERIOD_MILLIS = 2_000; //TODO make configurable
+    private static final int STREAM_RESET_DELAY_MILLIS = 1000; //TODO make configurable
+
+    private final HeartBeatServiceGrpc.HeartBeatServiceStub client;
+    private final ThreadPoolTaskScheduler taskScheduler;
+
+    private boolean isConnected;
+    private StreamObserver<AgentHeartBeat> requestObserver;
+    private ScheduledFuture<?> heartbeatFuture;
+    private String claimedJobId;
+    private AgentHeartBeat heartBeatMessage;
+
+    GrpcAgentHeartBeatServiceImpl(
+        final HeartBeatServiceGrpc.HeartBeatServiceStub client
+    ) {
+        this.client = client;
+        this.taskScheduler = new ThreadPoolTaskScheduler();
```

Comment:
This code relies on the task scheduler to schedule at most one task at the time.
i.e. only works if it's backed by a thread pool of max size 1.
The other reason for not wanting a shared one is that it may be used by other components for long-running tasks, whereas responsiveness is preferred here.
Any idea on managing these requirements on a shared, externally provided scheduler bean?

Reply:
Well I don't know about shared given those requirements. If you want to make it externally provided though you could use `@Qualified` on your bean creation to specify the name of the bean you actually want. Then you could have multiple `TaskSchedulers` configured and managed

Reply:
What's the value you see in doing that?
The only one i can think of is providing it as constructor allows mocking in tests, but in this case it's ok as is.